### DISCUSSION TITLE
source.on_leave(source.skip) produces Error 5: source has no method on_leave

### DISCUSSION THREAD

**Discussion by jascco** [savonet/liquidsoap#3914] [Q&A]
Here is a description of my radio station to put my question in context. I am also uploading the entire script which has no liquidsoap errors and is running behind the public/Internet facing website https://mynetradio.org

There are 16 music playlists. On weekdays there are two hour timeslots at 8, 10, 12, 14, 16, 18, 20, 22 hours and an 8 hour timeslot from midnight to 8 hours. On weekends one playlist goes 48 hours. A time driven switch chooses the playlist for the source music (lines 175-232)

The user can request any of the music playlists which will define the source userprogram (lines 265-307)

The user can request a song to be played from the current playlist which is pushed into the songRequests queue (lines 316-348)

A live broadcast can connect to input.harbor which defines the source live

An emergency single defines the source broken

Thus there are 5 possible sources which could be ready for the fallback (line 360)

streamme = fallback(id="streamme-fb", track_sensitive=false, [live, songRequests, userchoice, music, broken])

The problem is using track_sensitive=false causes an immediate switch which is what I want. Upon returning to a source it picks up in the middle of a track where it left instead of starting a new track.

Page 84 of the liquidsoap book talks about this problem and has two methods for solving it. One is to use fallback.skip and the example uses two sources.  If I understand the documentation it seems fallback.skip can only be used with two sources.

The other method used to work on my old station but now fails as there is no method on_leave for a playlist.

source.on_leave(source.skip) produces Error 5: source this value has no method on_leave

What is the recommended way to cause a source to skip when another source is selected by a switch?

Here is my liquidsoap script as a PDF with line numbers
[castle-alpine.pdf](https://github.com/savonet/liquidsoap/files/15284111/castle-alpine.pdf)

Here is the script as a .txt file
[castle-alpine.txt](https://github.com/savonet/liquidsoap/files/15284114/castle-alpine.txt)

**Comment by vitoyucepi** [comment]
Hi @jascco,
Here's a source for `fallback.skip`.
https://github.com/savonet/liquidsoap/blob/6acc8ab91a2d6057204be5eb5b702c39c39998ec/src/libs/extra/source.liq#L240-L260

As you can see, it uses the `fallback` function with two sources internally, so before using `fallback.skip` you can combine all the other sources into a single source.
```ruby
s1 = fallback([songRequests, userchoice, music, broken])
s2 = fallback.skip(live, fallback=s1)
```

As for `source.on_leave`, it was removed in #3096 for some reason. @toots didn't provide any migration instructions. Perhaps you could use `source.on_shutdown`.
```ruby
def skip_song()
  # Skip logic here  
end

source.on_shutdown(live, skip_song)
```

**Reply by jascco** [reply]
Hello @vitoyucepi,

Thank you for the suggestion for using fallback.skip by combining all the sources except one. I tested this by using:
```ruby
s1 = fallback([live, userchoice, music, broken])
s2 = fallback.skip(songRequests, fallback=s1)
```
This works when the user requests a song or songs. When the request queue is empty the fallback goes back to the userchoice or music source and plays a new song from the beginning. This is a partial solution. When the time driven switch goes to a new source then selecting the old source anytime later will play starting in the middle of a song.

I haven't tried source.on_shutdown, I need to read about it. Thank you for the suggestions.

I noticed in the liquidsoap log that it is decoding a song ahead on the current source. I've been wanting to display on my web site the next song that will play. I noticed that the telnet server has commands for source.next and returns the next song. I looked at the source for the telnet server web page and created an AJAX function to issue a command to the harbor for /telnet. This seems to work well but I don't know if there are any side effects or pitfalls when doing this. I am very new to using AJAX. Here is my AJAX script:

```js
function telnetCmd(command) {
	//alert('telnetCmd = ' + command);
	let xhr = new XMLHttpRequest;
	let url = new URL('http://wildoat05.local:8007/telnet');
	xhr.open('POST', url);
	xhr.responseType = 'text';
	xhr.send(command);
	xhr.onload = function() {
		if (xhr.status != 200) { // HTTP error
			document.getElementById("telnetResponse").innerHTML = xhr.status;
			return;
		}
		document.getElementById("telnetResponse").innerHTML = xhr.response;
	};
} //end of telnetCMD()
```

The function is called by telnetCmd('source.next')
The server wildoat05.local is only accessible from my local area network using an nginx virtual site. 

Thank you for any insights. I am new to using this site so if it's not proper etiquette to ask a different question in a reply rather than starting another new discussion, please let me know.

Jack

**Reply by vitoyucepi** [reply]
1. Why do you need to use telnet when you can use the [`http server`](https://www.liquidsoap.info/doc-2.2.4/harbor_http.html)?
1. Why do you use AJAX, it's pretty outdated js API? The browser has [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).
   Maybe even [ky](https://github.com/sindresorhus/ky) or [got](https://github.com/sindresorhus/got).
1. > from my local area network

   Why does this matter?

**Comment by jascco** [comment]
1. I'm sorry, I wasn't clear in my previous posting. I just completed converting from the old style shell exec "echo command | telnet localhost 8007" used in my old server to the harbor server where I've registered 8 functions. The example I posted is so I can use a couple of the commands available in the harbor web version of the telnet server not the actual telnet server itself. The particular command is source.next which returns the next song to be played by that source.
2. I didn't know about fetch - thank you for the suggestion, I will read about it. I wasn't sure my coding was going down a good path which is why I made the posting. I appreciate your guidance. I went the AJAX route as it was mentioned in the liquidsoap book on page 154. I looked at the source code to the harbor web page for the telnet server and it uses AJAX.
3. Less than a day after adding the ability to request playing a song from the displayed playlist, hackers in Singapore somehow found my site and using their automated probing tolls exploited a security hole in Wordpress to compromise my server. Thus I'm more paranoid about the control aspects. I've split the site into two - one a public/Internet site that is display only - no control aspect at all. The second site is private and currently only accessible from my local network while I'm developing it. The hackers still run automated probing but now I've made nginx very restrictive so all they get are 401, 403, and 404 errors. Since I'm a hobbyist and novice to much of this, I'm trying to be security conscious.

Thank you again for your suggestions and insight. I know there is a lot to learn.
